lib/improver/nbhood.py

```diff
@@ -228,6 +231,143 @@ def __repr__(self):
+        calculate the sum over a neighbourhood of size
+        (2*cells_x+1)*(2*cells_y+1). This sum is then divided by the area of
+        the neighbourhood to calculate the mean value in the neighbourhood.
+
+        At edge points, the sum and area of the neighbourhood are calculated
+        for each point individually. For non-edge points, a faster, vectorised
+        approach is taken:
+        1. The displacements between the four points used to calculate the
+           neighbourhood total sum and the central grid point are calculated.
+        2. Four copies of the cumulate array output are flattened and rolled
+           by these displacements to align the four terms used in the
+           neighbourhood total sum calculation.
+        3. The neighbourhood total at all non-edge points can then be
+           calculated simultaneously in a single vector sum.
+
+        Parameters
+        ----------
+        cube : iris.cube.Cube
+            Cube to which neighbourhood processing is being applied. Must
+            be passed through cumulate_array method first.
+        cells_x, cells_y : integer
+            The radius of the neighbourhood in grid points, in the x and y
+            directions (excluding the central grid point).
+
+        Returns
+        -------
+        cube : iris.cube.Cube
+            Cube to which square neighbourhood has been applied.
+        """
+        def _sum_and_area_for_edge_cases(cube, cells_x, cells_y, i, j):
+            """
+            Function to calculate the total sum and area of neighbourhoods
+            surrounding edge grid point (i,j) which can't use the flatten and
+            roll method. These values are used to calculate the mean value of
+            the neighbourhood.
+
+            Parameters
+            ----------
+            cube : iris.cube.Cube
+                Cube to which neighbourhood processing is being applied. Must
+                be passed through cumulate_array method first.
+            cells_x, cells_y : integer
+                The maximum radius of the neighbourhood in grid points, in the
+                x and y directions (excluding the central grid point). For
+                edge cases, the radius may be less than this, if the
+                neighbourhood falls off the domain edge.
+            i, j : integer
+                x and y indices of the grid point for which the total sum and
+                neighbourhood area is sought.
+
+            Returns
+            -------
+            total : float
+                The sum of all values in the neighbourhood surrounding grid
+                point (i,j).
+            area : integer
+                The area of the neighbourhood surrounding grid point (i,j).
+                This accounts for cases where the neighbourhood extends beyond
+                domain bounds.
+            """
+            x_min = i-cells_x-1
+            x_max = min(cube.shape[1]-1, i+cells_x)
```

Comment:
This is relying on your cube being in y, x order. Maybe you should be taking the length of the 'y' coordinate instead as you do below.